### DISCUSSION TITLE
hybrid_property with different attribute name cannot be referenced in SQLAlchemy 1.4

### DISCUSSION THREAD

**Discussion by RazerM** [sqlalchemy/sqlalchemy#9073] [Usage Questions]
I'm not sure if this counts as a regression since it's a bit weird, but if we have a hybrid_property factory function and the result is assigned to the class with a different name (see `namex = make_name()` line), then `row.name` works in SQLAlchemy 1.3 but not 1.4.

The reason we have a differently named property is backwards compatibility on our end, but it *seems* reasonable that the original hybrid_property's name should end up as the column name even if our class attribute is named differently.

```python
import sqlalchemy as sa
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import sessionmaker, scoped_session

Base = declarative_base()

engine = sa.create_engine('sqlite:///:memory:')

session_factory = sessionmaker()
session = scoped_session(session_factory)
session.configure(bind=engine)


def make_name():
    @hybrid_property
    def name(self):
        return self.firstname + " " + self.lastname

    return name


class User(Base):
    __tablename__ = "users"

    id = sa.Column(sa.Integer, primary_key=True)
    firstname = sa.Column(sa.String(20), nullable=False)
    lastname = sa.Column(sa.String(20), nullable=False)

    namex = make_name()


Base.metadata.create_all(engine)

user = User(firstname="John", lastname="Doe")
session.add(user)
session.flush()

query = session.query(User.id, User.namex)
a = query.first()

print("keys=", list(a.keys()))
```

SQLAlchemy 1.4:

```
keys= ['id']
```

SQLAlchemy 1.3:

```
keys= ['id', 'name']
```

The only issue I found which looks possibly related is https://github.com/sqlalchemy/sqlalchemy/issues/6718

**Comment by zzzeek** [comment]
so here's a way to fix that:

```py
def make_name(n):
    @hybrid_property
    def name(self):
        return self.firstname + " " + self.lastname

    name.__name__ = n
    return name

```

here's where it happens in the current code:

https://github.com/sqlalchemy/sqlalchemy/blob/a950402dae2a5b2448f5f4235946b2f767c7485c/lib/sqlalchemy/ext/hybrid.py#L1113-L1125


what can we do (later): perhaps we can figure out how to use [__set_name__()](https://docs.python.org/3/reference/datamodel.html#object.__set_name__) which is something fairly recent that is designed to help with this problem, however, even without that hook, declarative could be setting the "name" on the descriptor in any case.  The reason we don't is because we think (or maybe assumed / guessed ?) that some people were using the same hybrid descriptor in multiple places.   There might be an issue that discussed this (I think w.r.t. association proxy).   

OK the issue is likely #6215, that seems to be where we changed this (and considered it to be a bug in 1.3).    

I consider this to be a case we could still try to improve since the ``__set_name__()`` thing does seem to set a precedent for how descriptors are "supposed" to work.

**Reply by RazerM** [reply]
Thanks, that works for session.query. Any way to make it work for `connection.execute(select(...))` in this case:

```python
import sqlalchemy as sa
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import sessionmaker, scoped_session

Base = declarative_base()

engine = sa.create_engine('sqlite:///:memory:')

session_factory = sessionmaker()
session = scoped_session(session_factory)
session.configure(bind=engine)


def make_name():
    @hybrid_property
    def name(self):
        return self.firstname + " " + self.lastname

    @name.expression
    def name(cls):
        # This function doesn't matter, it just needs to be one
        return sa.func.coalesce(cls.firstname, cls.lastname)

    name.__name__ = 'namex'
    return name


class User(Base):
    __tablename__ = "users"

    id = sa.Column(sa.Integer, primary_key=True)
    firstname = sa.Column(sa.String(20), nullable=False)
    lastname = sa.Column(sa.String(20), nullable=False)

    namex = make_name()


Base.metadata.create_all(engine)

user = User(firstname="John", lastname="Doe")
session.add(user)
session.flush()

query = session.query(User.id, User.namex)
a = query.first()

print("keys=", list(a.keys()))

with engine.connect() as conn:
    result = conn.execute(sa.select(User.id, User.namex))
    print("keys=", result.keys())

```

output:

```
keys= ['id', 'namex']
keys= RMKeyView(['id', 'coalesce_1'])
```

**Reply by zzzeek** [reply]
oh well that's not new you need to apply a label() to your expression:

    return sa.func.coalesce(cls.firstname, cls.lastname).label("namex")

**Reply by RazerM** [reply]
That's true, I was trying to provide an alternative for a user that was doing the equivalent of the following where executing `query.statement` has the "right" keys on SQLAlchemy 1.3 but not 1.4.

```python
query = sa.orm.Query([User.id, User.namex])
with engine.connect() as conn:
    result = conn.execute(query.statement)
    print("keys=", result.keys())
```

I've created the above example based on how someone was reading a query into pandas:

```python
import pandas as pd
query = session.query(User.id, User.namex)
df = pd.read_sql(query.statement, session.connection())
print(df.columns)
```

My feeling is that sqlalchemy should be doing the executing and `row`/`row._mapping` used to construct a pandas DataFrame though.

**Reply by RazerM** [reply]
It also feels a bit weird to me that executing through a session has a different key

```
statement = sa.select(User.id, User.namex)
print('session keys', session.execute(statement).keys())
print('connection keys', session.connection().execute(statement).keys())
```
output:

```
session keys RMKeyView(['id', 'namex'])
connection keys RMKeyView(['id', 'coalesce_1'])
```

In SQLAlchemy 1.3 the behaviour difference was ORM query vs Core select but in SQLAlchemy 1.4 it's whether the connectable is a session or connection that makes the behaviour different.

**Reply by zzzeek** [reply]
they should send session to pd.read_sql() if they want it to have ORM features on the result set

**Reply by zzzeek** [reply]
if you do this:

```py
    @name.expression
    def name(cls):
        return sa.func.coalesce(cls.firstname, cls.lastname).label("namex")
```

then they both work

```py
with engine.connect() as conn:
    result = session.execute(query)
    print("keys=", result.keys())

    result = conn.execute(query)
    print("keys=", result.keys())

```

```
keys= RMKeyView(['id', 'namex'])
keys= RMKeyView(['id', 'namex'])

```

**Reply by RazerM** [reply]
pandas doesn't support sessions for some reason. 

I think that's the best solution, I'll combine it with the earlier `__name__` stuff:

```python
def make_name(n):
    @hybrid_property
    def name(self):
        return self.firstname + " " + self.lastname

    @name.expression
    def name(cls):
        return (cls.firstname + " " + cls.lastname).label(n)

    name.__name__ = n
    return name
```

**Reply by zzzeek** [reply]
> pandas doesn't support sessions for some reason.

my impression is that this is because pandas added a little bit of SQLAlchemy support in 2011 or something like that, then never looked again at it.   they should give it a revisit and update it for modern SQLAlchemy patterns.

> 
> I think that's the best solution, I'll combine it with the earlier **name** stuff:
> 
> ```python
> def make_name(n):
>     @hybrid_property
>     def name(self):
>         return self.firstname + " " + self.lastname
> 
>     @name.expression
>     def name(cls):
>         return (cls.firstname + " " + cls.lastname).label(n)
> 
>     name.__name__ = n
>     return name
> ```

yeah that will cover all the bases
